### PR TITLE
adding sharedservices

### PR DESCRIPTION
**Pull request by philbrookes**
creating or deleting a slice should check total number of slices in use, and update the "required_instances" in the SharedService object, if more or less are now required.

### CODE REVIEW

File: pkg/apis/aerogear/v1alpha1/types.go [hunk 1]
```diff
@@ -22,18 +28,66 @@ type SharedService struct {
 }
 
 type SharedServiceSpec struct {
+	Service                         string                 `json:"service"`
+	RequiredInstances               int                    `json:"required_instances"`
```
Comment (maleck13) [inline pkg/apis/aerogear/v1alpha1/types.go:32]: Not sure what this is for

File: pkg/operator/managed/handler.go [hunk 2]
```diff
@@ -101,16 +105,90 @@ func (h *Handler) getServiceClass(service *v1alpha1.SharedService) (*v1beta1.Clu
...
+	if err != nil {
+		return nil, err
+	}
+
+	si := buildServiceInstance(namespace, serviceClassExternalName, parameters, *sc)
+	return h.serviceCatalogClient.Servicecatalog().ServiceInstances(namespace).Create(&si)
 }
 
 func (h *Handler) handleSharedServiceCreateUpdate(service *v1alpha1.SharedService) error {
+	if len(service.Spec.CurrentInstances) < service.Spec.RequiredInstances {
```
Comment (maleck13) [inline pkg/operator/managed/handler.go:128]: why required instead of min

File: pkg/operator/managed/handler.go [hunk 3]
```diff
@@ -101,16 +105,90 @@ func (h *Handler) getServiceClass(service *v1alpha1.SharedService) (*v1beta1.Clu
...
+
+		SharedServiceInstance := &v1alpha1.SharedServiceInstance{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "aerogear.org/v1alpha1",
```
Comment (maleck13) [inline pkg/operator/managed/handler.go:132]: change to ```v1alpha1.SchemaGroup.String()```

File: pkg/operator/managed/handler.go [hunk 4]
```diff
@@ -101,16 +105,90 @@ func (h *Handler) getServiceClass(service *v1alpha1.SharedService) (*v1beta1.Clu
...
+			},
+		}
+
+		err := sdk.Create(SharedServiceInstance)
+		if err != nil {
+			fmt.Printf("error creating shared service instance: %v\n", err.Error())
+			return err
+		}
+		if service.Spec.CurrentInstances == nil {
+			service.Spec.CurrentInstances = []string{}
```
Comment (maleck13) [inline pkg/operator/managed/handler.go:153]: I wonder if for stuff like this should we have something like ```service.SetDefaults()````

**Review by maleck13 — APPROVED** [review]
Happy to merge as only poc code, not sure why we have required and min instances

File: pkg/operator/managed/handler.go (hunk 2 above)
Comment (philbrookes) [reply pkg/operator/managed/handler.go:128]: there is both required and min, I thought the idea would be that required is based on the number of used slices, but min is defined by the service (i.e. if the service takes a really long time to provision, don't ever go below 1, even if no slices are in use).

**Pull request merged**